docs/proposals/20220105-observatorium-alerting.md

```diff
@@ -0,0 +1,169 @@
+# Observatorium Alerting
+
+* **Owners:**
+  * `@onprem` `@bwplotka`
+
+* **Related Tickets:**
+  * https://issues.redhat.com/browse/MON-1761
+
+* **Other docs:*
+  * None 
+
+> TL;DR: For monitoring use cases, especially tenants of the metric signal of Observatorium, users want to be able to configure Prometheus-based alerts that will be evaluated for a given interval. If triggered they should notify the desired target (receiver) (e.g PagerDuty using Alertmanager Configuration). Users should also be able to see the status of all alerts through Prometheus-like alerts HTTP API. This proposal specifies how such a system can be deployed within Observatorium.
```

Comment:
TODO:
* How often do we reload AM conf
* How do we trigger reload?
* Does triggering reload affects AM
* Exact config file? (type)
* How tenancy is enabled?

Reply:
I believe the biggest question mark we have is still around multi-tenancy and Alertmanager configuration. I'll try to summarize the discussions so far.

#### Main concern:
- We do not have (yet) a good idea on how multi-tenancy with configuration should be implemented. The fact is that Alertmanager does not recognize tenancy, meaning the configuration, from global perspective, represents a configuration of a single user. To make tenancy possible, we need to design a solution which would makes it possible to isolate configurations of multiple users (tenants) while still being represented in the global configuration (on top of this, there are related questions such as how to enable such isolation when it comes to UI and other APIs not included in this proposal - i.e. silences, inhibitions - the design choices here might have impact on these as well).

#### Ideas:
1) Have separate Alertmanager instance for each tenant (hard tenancy)
- With this, we would completely side-step the issue of configuration, since the global config = user config. Downsides? Extra overhead of running Alertmanager instance for each tenant?

2) Combine configurations provided by users into a single global configuration (soft tenancy)
- This aligns with how we deal with e.g. rules (see [proposal](https://github.com/observatorium/observatorium/blob/322c22821b4c4b94a82186f45d7e1f2b4b3b3383/docs/proposals/20201019-prometheus-rules-in-observatorium-api.md?plain=1#L197)). This would mean creating one global file with each tenant being represented in the global configuration. The main issue with this solution is how to combine tenants configuration:
 - Alertmanager has a few global options that are applied if the child routes have no configuration set. If tenants are to provide us with their global parameters, it is not clear how it would be possible to combine global parameters from multiple tenants into one config file.
 - An additional suggestion how to solve this would be to accept only non-global parameters, i.e. `routes` and `receivers` and combine these into a global config file. The downside is obviously that users cannot leverage global options and would need to specify configuration for all routes explicitly. 

In this context it is also worth looking at how Cortex works, in which [Alertmanager ](https://cortexmetrics.io/docs/architecture/#alertmanager)is running as a component within Cortex. This makes it possible to run a 'multitenant Alertmanager' as a single component, which encompasses multiple Alertmanagers, each dedicated to a single tenant (meaning each tenant can provide their isolated, global configuration)